src/core/base_plugin_manager.h

```diff
@@ -0,0 +1,263 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __BASE_PLUGIN_MANAGER_H
+#define __BASE_PLUGIN_MANAGER_H
+
+#include "common/nixl_log.h"
+
+#include <string>
+#include <vector>
+#include <variant>
+#include <mutex>
+#include <filesystem>
+#include <memory>
+#include <functional>
+#include <map>
+
+/**
+ * Custom deleter for dlopen handles
+ */
+struct dlHandleDeleter {
+    std::string fini_func_name;
+
+    explicit dlHandleDeleter(std::string fini_name = "") : fini_func_name(std::move(fini_name)) {}
+
+    void
+    operator()(void *handle) const noexcept;
+};
+
+struct pluginConfig {
+    std::string initFuncName; // e.g., "nixl_plugin_init"
+    std::string finiFuncName; // e.g., "nixl_plugin_fini"
+    std::string filenamePrefix; // e.g., "libplugin_"
+    std::string filenameSuffix; // e.g., ".so"
+    int expectedApiVersion; // Expected API version for validation
+};
+
+template<typename pluginInterface> struct success_result {
+    std::unique_ptr<void, dlHandleDeleter> handle;
```

Comment:
This type occurs several times. Please consider adding an alias, e.g. `dl_handle_t`.